File: funk.c
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
void funksvd(int* pn, int* users, int* tracks, double* ratings, int* pnfeats, int* pnuser, int* pntrack, double* ufeats, double* tfeats) {
  int n = *pn;
  int nfeats = *pnfeats;
  int nuser = *pnuser;
  int ntrack = *pntrack;
  int min_loops = 30;
  double min_improve = 1e-4;
  double k = 4;
  double lrate = 1e-3;
  
  int i = 0;
  int j = 0;
  printf("Initialising user feats (%d x %d)\n", nuser, nfeats);
  for (i = 0; i < nuser * nfeats; i++) {
    //printf("Item %d\n", i);
    ufeats[i] = 1;
  }
  
  printf("Initialising track feats\n");
  for (i = 0; i < ntrack * nfeats; i++) {
    tfeats[i] = 1;
  }
  printf("Allocating cache\n");
  
  double* cached = calloc(n, sizeof(double));
  
  double sqerr = 0;
  double sqerr_old = 0;
  for (int feat = 0; feat < nfeats; feat++) {
    printf("Calculating feature %d (error = %f)\n", feat, sqrt(sqerr/n));
    int loops = 0;
    
    
    while ((loops < min_loops) || (sqrt(sqerr_old/n) - sqrt(sqerr/n) > min_improve)) {
      loops += 1;
      //printf("Iterating... (error = %f improvement = %f )\n", sqrt(sqerr/n), sqrt(sqerr_old/n) - sqrt(sqerr/n));
      sqerr_old = sqerr;
      sqerr = 0;
      for (i = 0; i < n; i++) {
        double prediction;
        int u = users[i];
        int t = tracks[i];
        prediction = cached[i] + ufeats[u + feat * nuser] * tfeats[t + feat * ntrack];
        double err = ratings[i] - prediction;
        double uv = ufeats[u + feat * nuser];
        ufeats[u + feat * nuser] += lrate * (err * tfeats[t + feat * ntrack] - ufeats[u + feat * nuser] * k);
        tfeats[t + feat * ntrack] += lrate * (err * uv - tfeats[t + feat * ntrack] * k);
        sqerr += err * err;
      }
    }
    for (i = 0; i < n; i++) {
      int u = users[i];
      int t = tracks[i];
      cached[i] += ufeats[u + feat * nuser] * tfeats[t + feat * ntrack];
    }
  }
  free(cached);
}

File: funk.R
funksvd <- function (users, tracks, ratings, nfeats=16) {
  n <- length(users);
  nuser <- max(users) + 1;
  ntrack <- max(tracks) + 1;
  ufeats <- double(nuser * nfeats);
  tfeats <- double(ntrack * nfeats);
  dyn.load('funk.so');
  sv <- .C('funksvd', as.integer(n), as.integer(users), as.integer(tracks), as.double(ratings), 
            as.integer(nfeats), as.integer(nuser), as.integer(ntrack), 
            ufeats=ufeats, tfeats=tfeats);
  ufeats <- matrix(sv$ufeats, nuser, nfeats);
  tfeats <- matrix(sv$tfeats, ntrack, nfeats);
  return(list(ufeats=ufeats, tfeats=tfeats));
  # baseline <- rep(mean(ratings), n);
  # cached <- baseline;
  # sqerr <- 0;
  # sqerr.old <- 0;
  # for (feat in 1:nfeats) {
  #   cat('Calculating feature', feat, '\n');
  #   loops <- 0;
  #   while (loops < min.loops | ((sqerr.old/n)^.5 - (sqerr/n)^.5 > min.improve)) {
  #     loops <- loops + 1;
  #     cat('Iterating... (error =', (sqerr/n)^.5, 'improvement =', (sqerr.old/n)^.5 - (sqerr/n)^.5, ')\n');
  #     sqerr.old <- sqerr;
  #     sqerr <- 0;
  #     prediction <- numeric(n);
  #     for (i in 1:n) {
  #       u <- users[i] + 1;
  #       t <- tracks[i] + 1;
  #       prediction <- cached[i] + t(ufeats[u,feat]) %*% tfeats[t,feat];
  #       err <- (ratings[i] - prediction);
  #       sq <- err^2;
  #       uv <- ufeats[u,feat];
  #       ufeats[u,feat] <- ufeats[u,feat] * (1 - k * lrate) + lrate * err * tfeats[t,feat];
  #       tfeats[t,feat] <- tfeats[t,feat] * (1 - k * lrate) + lrate * err * uv;
  #       sqerr <- sqerr + sq;
  #     }
  #   }
  #   for (i in 1:n) {
  #     u <- users[i] + 1;
  #     t <- tracks[i] + 1;
  #     cached[i] <- cached[i] + t(ufeats[u,feat]) %*% tfeats[t,feat];
  #   }
  # }
  # return(list(baseline=baseline,ufeats=ufeats, tfeats=tfeats));
}

funkpred <- function(train, test, ratings) {
  n <- nrow(train);
  ntrack <- max(train$Track) + 1;
  nuser <- max(train$User) + 1;
  mu <- mean(ratings);
  baseline <- rep(mu, n);

  usum <- numeric(nuser);
  ucount <- numeric(nuser);
  for (i in 1:n) {
    u <- train$User[i] + 1;
    usum[u] <- usum[u] + ratings[i] - baseline[i];
    ucount[u] <- ucount[u] + 1;
  }
  umean <- usum / (ucount + 5);
  baseline <- baseline + umean[train$User + 1];
  
  tsum <- numeric(ntrack);
  tcount <- numeric(ntrack);
  for (i in 1:n) {
      t <- train$Track[i] + 1;
      tsum[t] <- tsum[t] + ratings[i] - baseline[i];
      tcount[t] <- tcount[t] + 1;
    }
  tmean <- tsum / (tcount + 25);
  baseline <- baseline + tmean[train$Track + 1];



  sv <- funksvd(train$User, train$Track, ratings - baseline, 64);
  pred <- sapply(1:nrow(test), 
    function (i) {
      trk <- test$Track[i] + 1;
      usr <- test$User[i] + 1;
      base <- umean[usr] + tmean[trk] + mu;
      s <- t(sv$ufeats[usr,]) %*% sv$tfeats[trk,];
      return(base + s);
    });
  pred.train <- sapply(1:nrow(train), 
    function (i) {
      trk <- train$Track[i] + 1;
      usr <- train$User[i] + 1;
      base <- umean[usr] + tmean[trk] + mu;
      s <- t(sv$ufeats[usr,]) %*% sv$tfeats[trk,];
      return(base + s);
    });
  cat("Sanity check RMSE:", rmse(pred.train, ratings), '\n');
  return(list(pred=pred));
}